### DISCUSSION TITLE
Network Storage: Addressing data loss and recovering from network mount failures gracefully

### DISCUSSION THREAD

**Discussion by jmealo** [home-assistant/architecture#1033] [General]
# Proposed fix for: https://github.com/home-assistant/core/issues/106446

### When writing a backup

- Use `findmnt` or similar to make sure that the network mount is in a good state when creating/writing a backup.
- **If the mount is in a good state:** proceed.
- **If the mount is in a bad state, and a fallback location is configured:** write to the fallback location and raise a notice to the user that the last backup was written to the fallback location.
- **If the mount is in a bad state, and a fallback location is not configured:** raise a notice to the user that the last backup failed to write.

### Follow-up enhancement
- Add option to copy backups from the fallback location when the mount is restored.

## Recovering from the error condition
- In the event that the mount point is a regular directory and contains valid backups, it should become the fallback directory, and a notice should be raised to the user that this has occurred.

## Follow-up enhancement
- Raise notice that backups are present in the fallback location but the option to copy them automatically has not been set up.

# UI Fixes
- Show fallback mount configuration guidance when network storage is the default destination and the mount fails (instead of showing the blank page). Explain how to setup robust backups with a transient storage server.

### Steps to reproduce the issue

1. Use network storage for backups
2. Have a backup occur when the network storage is unavailable
3. The next time the mount reconnects, it will fail due to a non-empty directory

**Comment by 0sy** [comment]
Hopefully this will be addressed soon. No backups for a week, I feel naked.

**Comment by chriszuercher** [comment]
In my opinion, there should be another step between

- Use findmnt or similar to make sure that the network mount is in a good state when creating/writing a backup.
- If the mount is in a good state: proceed.

which should be:
- Try to remount the failed network mount

As in my case, the NAS (Synology) is most probably not running when Hassio starts, as I turn it on only for backup tasks and is shutdown afterward to save energy. So the network share cannot be mounted on the start of Hassio, but could be when the backup starts (as the NAS is running then).

**Comment by member1980** [comment]
same here....hope there will be a fix

**Comment by agners** [comment]
This is more or less the same as in https://github.com/home-assistant/supervisor/issues/4856. Since the issue is marked as resolved, I am close this as well.

**Reply by chriszuercher** [reply]
@agners is automatic "try remounting the share" also part of https://github.com/home-assistant/supervisor/issues/4856 ? Or is there a way this could be triggered somehow like in an servicecall inside an automation?

**Reply by member1980** [reply]
@agners Before the latest update the issue has been automatically fixed, after the NAS got back online. Since the latest update, i have to click manually on "fix now". Only after this manual step the backup routine is able to use NAS again. This "fix" is useless, wenn i have to manually "fix" it every day. Don't you think?!

**Reply by agners** [reply]
> @agners is automatic "try remounting the share" also part of https://github.com/home-assistant/supervisor/issues/4856 ? Or is there a way this could be triggered somehow like in an servicecall inside an automation?

No, that is not implemented (yet). Currently, when the mount is down at the time the backup starts, an error is raised.

> Before the latest update the issue has been automatically fixed, after the NAS got back online. Since the latest update, i have to click manually on "fix now". 

Unfortunately, this architecture discussion here doesn't exactly describe the error which should be addressed, but from my point of view the main problem was that sometimes the mount did not got setup correcltly, and backups landed on the local disk instead of the remote mount. Also reloading the mount in this cases did not work, even though the server was back online. This should be addressed with https://github.com/home-assistant/supervisor/pull/4872.

In general, network storage (mounts) are largely handled by the Linux kernel itself. If the network storage was reachable when you started Home Assistant (or rather Supervisor specifically), then the mount is being setup. From that point onwards, it is on Linux to retry/reconnect in case the server goes down. As long as you don't restart the system, the mount will remain active and useable.

This remains the same. However, we do have more situations where Supervisor detects when the server went down, and hence raises the repair issue. However, the mount on a Linux kernel level is still active at that point. It is just not usable as long as the server is not online (obviously). But it should become usable as soon as the server is back online. I've tested this behavior on my end: In your use case, you essentially can just ignore the repair issue and things should continue to work as before.

Now this is not proper automatic remount implementation yet, as when you restart your Home Assistant Server when the NAS is down, the mount will not get setup. And as mentioned initially, in that case the Supervisor does not setup the mount on it's own.

Essentially, from my point of view, to fullfill the use cases you describe there are two issues remaining:

- Mount should get setup automatically (in case the NAS was offline when the OS/Supervisor started)
- The repair should automatically clear once the mount comes available again (this is more cosmetic)

**Reply by member1980** [reply]
Thanks for you response. I think i will switch my backup storage for HA to my router which is permanently online. ;-)

**Reply by chriszuercher** [reply]
Thanks for clarification. You're right. For my usecase (NAS offline during Hassio Boot) i would need a way to automatically remount the share (as soon the NAS is online). This would also fix the second point I guess?

Couldn't the manual Fix using the button not just be automated to try this in a regular interval (like a cronjob?)

**Comment by 0sy** [comment]
This is sorted for me, no more issues

**Comment by apachelance** [comment]
I am having the same issue: HA is running 24/7 on a small server. The NAS is only running a few hours a day. The NAS storage is mounted as backup destination but comes unavailable if the NAS turns off. And I have to fix it manually. This way the automatic backups are not done on the NAS but locally. Is there a fix work in progress? Or any other workaround via automation in terms of reconnection the NAS storage if it is online again?

Thanks.

**Comment by zoechi** [comment]
I still have this issue with 2024.12.4
I just occassionally restart my samba server but HomeAssistant never reconnects after the server comes back and the Reload offered by the repair notifications rarely helps.
The only ways I found to get the connection back is to remove and re-add the storage or reboot the server.

**Comment by Glideh** [comment]
Still having it on 2025.2.1
The issue remains showed
![image](https://github.com/user-attachments/assets/f0bd5f32-1842-441d-9558-7adb594c4930)
And not fixable without restarting HA
![image](https://github.com/user-attachments/assets/607f49a0-d025-41d6-a4a9-a6b6e5168e6a)

Backups are however well done on the remote device.

**Reply by frenck** [reply]
👋 Hi there @jmealo!

This is not an issue tracker, if you experience issues, please raise it in the appropriate issue tracker. If you unsure where to go, please check:

https://www.home-assistant.io/help

Thanks! 👍 

../Frenck
